keps/sig-storage/20200609-csi-service-account-token.md

```diff
@@ -0,0 +1,266 @@
+    Storage provider is expected to call TokenReview api where audience is
+    specified to validate the token.
+2.  If storage provider requires periodic authentication, `RequiresRemount` can
+    be used to trigger a sequence of `NodePublishVolume` calls. Each call will
+    have valid token in the `VolumeContext`.
+
+### Example Workflow
+
+Take the Vault provider for secret store CSI driver as an example:
+
+1.  Create `CSIDriver` object with `ServiceAccountTokenAudiences=['vault']` and
+    `RequiresRemount=true`
+2.  When the volume manager of kubelet sees a new volume, the pod object in
+    `mountedPods` will have `requiresRemound=true` after `MarkRemountRequired`
+    is called. `MarkRemountRequired` will call into `RequiresRemount` of the
+    in-tree csi plugin to fetch the `CSIDriver` object.
+3.  Before `NodePublishVolume` call, kubelet will request token from
+    `TokenRequest` api with `audiences=['vault']`.
+4.  The token will be specified in `VolumeContext` to `NodePublishVolume` call.
+5.  Every 0.1 second, the reconciler component of volume manager will remount
+    the volume in case the vault secrets expire and re-login is required.
+
+### Notes/Constraints/Caveats
+
+The `RequiresRemount` is useful when the mounted volumes can expire and the
+availability and validity of volumes are continuously required. Those volumes
+are most likely credentials which rotates for the best security practice. There
+are two options when the remount failed:
+
+1.  Keep the container/pod running and use the old credentials.
+    - The next `NodePublishVolume` may succeed if it was unlucky transient
+      failure.
+    - Given there are multiple of 0.1 second usage of stale credentials, it is
+      critical for the credential provisioners to guarantee that the validity
+      is revoked after expiry. In general, it is much harder to eliminate the
+      sinks than source.
+    - The container/pod will also have better observability in usage of the
+      stale credentials.
+2.  Kill the container/pod and hopefully the new container/pod has the refreshed
+    credentials.
+    - This will reduce the stale volume exposure by one sink.
+    - More likely to overcome fatal errors.
+    - Constart-up cost is high
+
+Option 1 is adopted. See discussion
+[here](https://github.com/kubernetes/enhancements/pull/1855#discussion_r443040359).
+
+### Test Plan
+
+- Unit tests around all the added logic in kubelet.
+- E2E tests around remount and token passing.
+
+### Graduation Criteria
+
+#### Alpha->Beta
+
+- Implemented the feature.
+- Wrote all the unit and E2E tests.
+
+#### Beta->GA
+
+- Deployed the feature in production and went through at least minor k8s
+  version.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  - Feature gate name: CSIDriverServiceAccountToken
+  - Components depending on the feature gate: kubelet
```

Comment:
Also apiserver, which means the next answer is yes.

Reply:
you are right